go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,271 @@
+	defer e.lock.Unlock()
+	for _, dnsName := range dnsNames {
+		if _, exists := e.dnsEntries[dnsName]; !exists {
+			e.dnsEntries[dnsName] = &dnsEntry{
+				namespaces: map[string]struct{}{
+					namespace: {},
+				},
+			}
+			e.signalAdded(dnsNamespace{dnsName: dnsName, namespace: namespace})
+		} else {
+			// only need to add the namespace to the struct
+			e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+		}
+	}
+	return nil
+
+}
+
+func (e *EgressDNS) updateEntryForName(dnsNamespace dnsNamespace) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	ips := e.dns.GetIPs(dnsNamespace.dnsName)
+	//TODO: if the old and new resolutions are the same should not update the dnsObject
+	e.dnsEntries[dnsNamespace.dnsName].dnsResolves = ips
+
+	//update the dnsObject
+	dnsObject, err := e.wf.GetDNSObject(e.nodeName)
+	existed := true
+	if errors.IsNotFound(err) {
+		// the object *MOST LIKELY* has not been created yet
+		dnsObject = &dnsobjectapi.DNSObject{
+			ObjectMeta: metav1.ObjectMeta{Name: e.nodeName},
+			Spec:       dnsobjectapi.DNSObjectSpec{},
+		}
+		existed = false
+	} else if err != nil {
+		return err
+	}
+	var ipStrings []string
+	for _, ip := range e.dnsEntries[dnsNamespace.dnsName].dnsResolves {
+		ipStrings = append(ipStrings, ip.String())
+	}
+
+	if dnsObject.Spec.DNSObjectEntries == nil {
+		dnsObject.Spec.DNSObjectEntries = make(map[string]dnsobject.DNSObjectEntry)
+	}
+
+	dnsObjectEntry := dnsObject.Spec.DNSObjectEntries[dnsNamespace.dnsName]
+	dnsObjectEntry.IPAddresses = ipStrings
+	dnsObject.Spec.DNSObjectEntries[dnsNamespace.dnsName] = dnsObjectEntry
+
+	if existed {
+		err = e.k.UpdateDNSObject(dnsObject)
+	} else {
+		_, err = e.k.CreateDNSObject(dnsObject)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	return e.dns.Update(dns)
+}
+
+// Run spawns a goroutine that handles updates to the dns entries for dnsNames used in
+// EgressFirewalls. The loop runs after receiving one of two signals
+// 1. a new dnsName has been added and a signal is sent to add the new DNS name, if an
+//    EgressFirewall uses a DNS name already added by another egressFirewall the previous
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsNamespace dnsNamespace
+	var ttl time.Time
+	var timeSet bool
+	// initially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsNamespace := <-e.added:
+				if err := e.dns.Add(dnsNamespace.dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsNamespace); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
+				if len(dnsNamespace.dnsName) > 0 {
```

Comment:
Not familiar with the code but, after the timeout we only update one entry of the map here.

Reply:
isn't dnsNamespace nil here?

Reply:
I had the same initial impression, but it gets re-initialized in line 185

Reply:
@squeed @fedepaol the way this works is that there is only one dnsName that ever gets updated at one time. the dnsName with the lowest TTL.  After the 30 minute timeout the goroutine wakes up checks if the lowest TTL needs to be updated and if not goes back to sleep